Include/pyport.h

```diff
@@ -470,6 +470,13 @@ extern "C" {
 #  define WITH_THREAD
 #endif
 
+/* Some WebAssembly platforms do not provide a working pthread implementation.
+ * Thread support is stubbed and any attempt to create a new thread fails.
+ */
+#if !defined(__wasi__) && (!defined(__EMSCRIPTEN__) || defined(__EMSCRIPTEN_PTHREADS__))
+#  define Py_CAN_START_THREADS 1
+#endif
```

Comment:
Unfortunately it's trickier than that because you _can_ do a WASI build with experimental threading support (the "experimental" bit is on the WASI side, not our side); see the `--with-wasm-pthreads` for turning the support on.

https://github.com/python/cpython/blob/6f09f69b7f85962f66d10637c3325bbb2b2d9853/configure.ac#L2123-L2127

Reply:
@brettcannon, do you think the general approach here is a reasonable? Previously, I had been just checking for thread support on the Python side, but that's a bit awkward for these C API tests.

I'm a bit confused by your comment that you can do a **WASI** build with threading support. I see that there is Emscripten support for pthreads. That's already handled by the `__EMSCRIPTEN_PTHREADS__` check on line 476.

The Python equivalent assumes that **WASI** never supports threads:
https://github.com/python/cpython/blob/6f09f69b7f85962f66d10637c3325bbb2b2d9853/Lib/test/support/threading_helper.py#L232-L233

I see that there is a WASI proposal for threads (https://github.com/WebAssembly/wasi-threads), but that looks like it's still marked "phase 1".

Reply:
> The Python equivalent assumes that **WASI** never supports threads

The threading support isn't well-tested because I haven't set up a buildbot yet and made it priority to more thoroughly test it. 😅

> I see that there is a WASI proposal for threads (https://github.com/WebAssembly/wasi-threads), but that looks like it's still marked "phase 1".

Yep, and it's staying at that phase because threading is being taken up by the WebAssembly CG at the W3C which acts as an upstream to WASI. That proposal hit I think phase 4 last month (it's definitely moving forward regardless of whether I got the number right), hence why this whole "threads in WASI" thing continues to be experimental both in WASI and for us.

I think we need to make a decision about free-threading and WASI. Do we want to try and make it work with the experimental threading support that's there, or do we say it's not worth it and wait on the more official support to land?

Reply:
👍 if the `HAVE_PTHREADS_STUBS` solution works.

Reply:
Hmmm... `HAVE_PTHREAD_STUBS` did not seem to work.

Reply:
It appears that when built without thread support WASI uses `HAVE_PTHREAD_STUBS`, but Emscripten stubs pthreads internally.